-for loops iterate over every element in a specific sequence

-while loops iterate until some confition end them

-Sometimes we need more control over when a loop should end, or skip an iteration. In these cases, we use the break and continue keywords, which can be used in both for and while loops.

*break ---> terminates a loop

*continue ---> skips one iteration of a loop.

Debugging ---> is the process of identifying and removing errors in the code

In [7]:
#Assume we have some products and we want to full a chip with them, but the weight of them should not exceed 100 kg
print ("First Method")
manifest = [("bananas", 15), ("mattresses", 24), ("dog kennels", 42), ("machine", 120), ("cheeses", 5)]
weight = 0
items = [ ]
for item_name, item_weight in manifest:
  print("current weight: {}".format(weight))
  if weight >= 100:
    break
  else:
      print("  adding {} ({})".format(item_name, item_weight))
      items.append(item_name)
      weight += item_weight
print("\nFinal Weight: {}".format(weight))
print("Final Items: {}".format(items))

First Method
current weight: 0
  adding bananas (15)
current weight: 15
  adding mattresses (24)
current weight: 39
  adding dog kennels (42)
current weight: 81
  adding machine (120)
current weight: 201

Final Weight: 201
Final Items: ['bananas', 'mattresses', 'dog kennels', 'machine']


-In the last code, we can notice that the final weight exceeded 100 kg

-We can see that the summation statement executed before check condition..Here when the weight was 81 and re-checked by the condition-->the result was true as 81<100, then it is summed again with the final item's weight so the final weight became 201, so that the result of the condition was false, so the loop was terminated.... so we reach fot that the output exceeds 100 as the summation statement is alwayes executed before the check's condition...

**To solve this problem we did debugging by adding print function after every statement.

-so we need to add another check condition for the final weight after the first condition, we will use continue keyword in this condition as we want to skip the last of loop and print the final weight..

---note that, we used continue not break to make the loop skip only this condition but continue the iteration to the last element in the list. May the next element's weight plus the preprevious item's weight doesn't exceed 100. As we will see in the following code:

In [10]:
print("\n Second Method")
manifest = [("bananas", 15), ("mattresses", 24), ("dog kennels", 42), ("machine", 120), ("cheeses", 5)]
weight = 0
items = [ ]
for item_name, item_weight in manifest:
  print("current weight: {}".format(weight))
  if weight >= 100:
    print("  breaking from the loop now!")
    break
  elif weight + item_weight > 100:
    print("  skipping {} ({})".format(item_name, item_weight))
    continue
  else:
    print("  adding {} ({})".format(item_name, item_weight))
    items.append(item_name)
    weight += item_weight
print("\nFinal Weight: {}".format(weight))
print("Final Items: {}".format(items))


 Second Method
current weight: 0
  adding bananas (15)
current weight: 15
  adding mattresses (24)
current weight: 39
  adding dog kennels (42)
current weight: 81
  skipping machine (120)
current weight: 81
  adding cheeses (5)

Final Weight: 86
Final Items: ['bananas', 'mattresses', 'dog kennels', 'cheeses']


Write a loop with a break statement to create a string, news_ticker, that is exactly 140 characters long. You should create the news ticker by adding headlines from the headlines list, inserting a space in between each headline. If necessary, truncate the last headline in the middle so that news_ticker is exactly 140 characters long.

Remember that break works in both for and while loops. Use whichever loop seems most appropriate. Consider adding print statements to your code to help you resolve bugs.

In [2]:
# HINT: modify the headlines list to verify your loop works with different inputs
headlines = ["Local Bear Eaten by Man",
             "Legislature Announces New Laws",
             "Peasant Discovers Violence Inherent in System",
             "Cat Rescues Fireman Stuck in Tree",
             "Brave Knight Runs Away",
             "Papperbok Review: Totally Triffic"]

news_ticker = ""
# write your loop here
for headline in headlines:
    news_ticker += headline + " "
    if len(news_ticker) >= 140:
        news_ticker = news_ticker[:140]
        break

print(news_ticker)
print(len(news_ticker))

Local Bear Eaten by Man Legislature Announces New Laws Peasant Discovers Violence Inherent in System Cat Rescues Fireman Stuck in Tree Brave
140


Prime numbers are whole numbers that have only two factors: 1 and the number itself. The first few prime numbers are 2, 3, 5, 7.

For instance, 6 has four factors: 1, 2, 3, 6.
1 X 6 = 6
2 X 3 = 6
So we know 6 is not a prime number.

In the following coding environment, write code to check if the numbers provided in the list check_prime are prime numbers.

If the numbers are prime, the code should print "[number] is a prime number."
If the number is NOT a prime number, it should print "[number] is not a prime number", and a factor of that number, other than 1 and the number itself: "[factor] is a factor of [number]".

In [4]:
check_prime = [26, 39, 51, 53, 57, 79, 85]
for number in check_prime:
    for i in range(2,number):
        if (number % i) == 0:
            print("{} is not a prime number, because {} is a factor of {}".format(number, i, number))
            break
        if i == number - 1:
            print("{} is a prime number".format(number))

26 is not a prime number, because 2 is a factor of 26
39 is not a prime number, because 3 is a factor of 39
51 is not a prime number, because 3 is a factor of 51
53 is a prime number
57 is not a prime number, because 3 is a factor of 57
79 is a prime number
85 is not a prime number, because 5 is a factor of 85
